assembly/src/parsers/io_ops.rs

```diff
@@ -1,32 +1,47 @@
 use super::{
-    parse_checked_param, parse_decimal_param, parse_element_param, parse_hex_param, parse_param,
-    Felt,
+    get_valid_felt, parse_checked_param, parse_decimal_param, parse_hex_or_decimal_param,
+    parse_hex_param, parse_param, Felt,
     Instruction::*,
     Node::{self, Instruction},
     ParsingError, Token, Vec,
 };
-use crate::{validate_operation, ADVICE_READ_LIMIT, MAX_PUSH_INPUTS};
+use crate::{validate_operation, StarkField, ADVICE_READ_LIMIT, HEX_CHUNK_SIZE, MAX_PUSH_INPUTS};
 
 // CONSTANTS
 // ================================================================================================
 
-/// The required length of the hexadecimal representation for an input value when more than one hex
-/// input is provided to `push` without period separators.
-const HEX_CHUNK_SIZE: usize = 16;
+/// The maximum parts number allowed for the `push` instruction.
+const MAX_PUSH_PARTS: usize = MAX_PUSH_INPUTS + 1;
 
 // INSTRUCTION PARSERS
 // ================================================================================================
 
-/// Returns `PushConstants` instruction node.
+/// Returns one of the `Push` instruction nodes.
 ///
 /// # Errors
 /// Returns an error if the instruction token has invalid values or inappropriate number of
 /// values.
 pub fn parse_push(op: &Token) -> Result<Node, ParsingError> {
-    validate_operation!(op, "push", 1..MAX_PUSH_INPUTS);
-
-    let constants = parse_constants(op)?;
-    Ok(Instruction(PushConstants(constants)))
+    debug_assert_eq!(op.parts()[0], "push");
+    match op.num_parts() {
+        0 => unreachable!(),
+        1 => Err(ParsingError::missing_param(op)),
+        2 => {
+            let param_str = op.parts()[1];
+            if let Some(param_str) = param_str.strip_prefix("0x") {
+                // if we have only one parameter
+                if param_str.len() <= HEX_CHUNK_SIZE {
+                    parse_push_one(op, ParamType::Hex, param_str)
+                } else {
+                    parse_push_many(op, ParamType::Hex, param_str)
+                }
+            } else {
+                parse_push_one(op, ParamType::Decimal, param_str)
+            }
+        }
+        3..=MAX_PUSH_PARTS => parse_push_many(op, ParamType::HexAndDecimal, ""),
```

Comment:
Instead of using the same `parse_push_many()` in different ways for different situations, I'd split it in two:

1. Two separate functions which parse the parameters into a vector of `u64` values. You already have these: `parse_constants()` and `parse_hex_params()`.
2. One common function which converts a vector of `u64` values into an appropriate instruction. This would be very similar to `parse_push_many()`, but since it would no longer be responsible for parsing, I'd call it `build_push_many_instruction()` or something like that.

The above would also allow you to get rid of `ParamType` enum.